cpp/src/arrow/engine/substrait/options.cc

```diff
@@ -0,0 +1,99 @@
+#include "arrow/engine/substrait/expression_internal.h"
+#include "arrow/engine/substrait/relation_internal.h"
+#include "substrait/extension_rels.pb.h"
+
+namespace arrow {
+namespace engine {
+
+class DefaultExtensionProvider : public ExtensionProvider {
+ public:
+  Result<DeclarationInfo> MakeRel(const std::vector<DeclarationInfo>& inputs,
+                                  const google::protobuf::Any& rel,
+                                  const ExtensionSet& ext_set) override {
+    if (rel.Is<arrow::substrait::AsOfJoinRel>()) {
+      arrow::substrait::AsOfJoinRel as_of_join_rel;
+      rel.UnpackTo(&as_of_join_rel);
+      return MakeAsOfJoinRel(inputs, as_of_join_rel, ext_set);
+    }
+    return Status::NotImplemented("Unrecognized extension in Susbstrait plan: ",
+                                  rel.DebugString());
+  }
+
+ private:
+  Result<DeclarationInfo> MakeAsOfJoinRel(
+      const std::vector<DeclarationInfo>& inputs,
+      const arrow::substrait::AsOfJoinRel& as_of_join_rel, const ExtensionSet& ext_set) {
+    if (inputs.size() < 2) {
+      return Status::Invalid("substrait::AsOfJoinNode too few input tables: ",
+                             inputs.size());
+    }
+    // on-key
+    if (!as_of_join_rel.has_on()) {
+      return Status::Invalid("substrait::AsOfJoinNode missing on-key");
+    }
+    ARROW_ASSIGN_OR_RAISE(auto on_key_expr, FromProto(as_of_join_rel.on(), ext_set, {}));
+    if (on_key_expr.field_ref() == NULLPTR) {
+      return Status::NotImplemented("substrait::AsOfJoinNode non-field-ref on-key");
+    }
+    const FieldRef& on_key = *on_key_expr.field_ref();
+
+    // by-key
+    std::vector<FieldRef> by_key;
+    for (const auto& by_item : as_of_join_rel.by()) {
+      ARROW_ASSIGN_OR_RAISE(auto by_key_expr, FromProto(by_item, ext_set, {}));
+      if (by_key_expr.field_ref() == NULLPTR) {
+        return Status::NotImplemented("substrait::AsOfJoinNode non-field-ref by-key");
+      }
+      by_key.push_back(*by_key_expr.field_ref());
+    }
+
+    // schema
+    int64_t tolerance = as_of_join_rel.tolerance();
+    std::vector<std::shared_ptr<Schema>> input_schema(inputs.size());
+    for (size_t i = 0; i < inputs.size(); i++) {
+      input_schema[i] = inputs[i].output_schema;
+    }
+    ARROW_ASSIGN_OR_RAISE(
+        auto schema, compute::asofjoin::MakeOutputSchema(input_schema, on_key, by_key));
+    compute::AsofJoinNodeOptions asofjoin_node_opts{std::move(on_key), std::move(by_key),
+                                                    tolerance};
+
+    // declaration
+    std::vector<compute::Declaration::Input> input_decls(inputs.size());
+    for (size_t i = 0; i < inputs.size(); i++) {
+      input_decls[i] = inputs[i].declaration;
+    }
+    return DeclarationInfo{
+        compute::Declaration("asofjoin", input_decls, std::move(asofjoin_node_opts)),
+        std::move(schema)};
+  }
+};
+
+std::shared_ptr<ExtensionProvider> ExtensionProvider::kDefaultExtensionProvider =
```

Comment:
Should we move this to substrait/options.h file? (Since the `kDefaultNamedTableProvider` is also there)